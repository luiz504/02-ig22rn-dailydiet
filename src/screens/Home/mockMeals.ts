import { Statistics } from '~/models/Statistics'
import { DailyMeals } from './types'

export const mockMeals: DailyMeals = {
  '12.08.22': [
    {
      id: 'i0',
      name: 'x-tuda0 1hhui 1hui2 12h3uih1 iu2hh13uih23h 12hh1h2uh3uh1i2h3ui1h2ui3hi',
      description:
        'Delicious 1hhui 1hui2 12h3uih1 iu2hh13uih23h 12hh1h2uh3uh1i2h3ui1h2ui3hi',
      date: new Date().toISOString(),
      inDiet: false,
    },
    {
      id: 'i1',
      name: 'x-tudim',
      description: 'Delicious',
      date: new Date().toISOString(),
      inDiet: false,
    },
    {
      id: 'i2',
      name: 'salad',
      description: '500g mix tropical salad',
      date: new Date().toISOString(),
      inDiet: true,
    },
  ],
  '13.08.22': [
    {
      id: 'i3',
      name: 'x-tudim',
      description: 'Delicious',
      date: new Date().toISOString(),
      inDiet: false,
    },
    {
      id: 'i4',
      name: 'salad',
      description: '500g mix tropical salad',
      date: new Date().toISOString(),
      inDiet: true,
    },
  ],
  '13.08.20': [
    {
      id: 'i3',
      name: 'x-tudim',
      description: 'Delicious',
      date: new Date().toISOString(),
      inDiet: false,
    },
    {
      id: 'i4',
      name: 'salad',
      description: '500g mix tropical salad',
      date: new Date().toISOString(),
      inDiet: true,
    },
  ],
  '13.08.21': [
    {
      id: 'i3',
      name: 'x-tudim',
      description: 'Delicious',
      date: new Date().toISOString(),
      inDiet: false,
    },
    {
      id: 'i4',
      name: 'salad',
      description: '500g mix tropical salad',
      date: new Date().toISOString(),
      inDiet: true,
    },
  ],
  '13.08.25': [
    {
      id: 'i3',
      name: 'x-tudim',
      description: 'Delicious',
      date: new Date().toISOString(),
      inDiet: false,
    },
    {
      id: 'i4',
      name: 'salad',
      description: '500g mix tropical salad',
      date: new Date().toISOString(),
      inDiet: true,
    },
  ],
  '13.08.23': [
    {
      id: 'i3',
      name: 'x-tudim',
      description: 'Delicious',
      date: new Date().toISOString(),
      inDiet: false,
    },
    {
      id: 'i4',
      name: 'salad',
      description: '500g mix tropical salad',
      date: new Date().toISOString(),
      inDiet: true,
    },
  ],
}

export const mockStatistics: Statistics = {
  bestInDietSequence: 22,
  inDietEntries: 99,
  outDietEntries: 10,
  totalEntries: 99 + 10,
  inDietPercentage: (99 / (99 + 10)) * 100,
}
