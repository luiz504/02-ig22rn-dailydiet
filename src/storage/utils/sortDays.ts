export const sortDays = (array: string[]) => {
  return array.sort((a, b) => {
    return new Date(b).getTime() - new Date(a).getTime()
  })
}
