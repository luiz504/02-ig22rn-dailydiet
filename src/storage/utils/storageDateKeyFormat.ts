import { format } from 'date-fns'

export function storageDateKeyFormat(date: Date): string {
  return format(date, 'yyyy/MM/dd')
}
