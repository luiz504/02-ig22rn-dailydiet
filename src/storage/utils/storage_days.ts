import AsyncStorage from '@react-native-async-storage/async-storage'
import { DAY_COLLECTION } from '../commons'

export async function getStoredDays() {
  return AsyncStorage.getItem(DAY_COLLECTION).then((storedData) =>
    storedData ? (JSON.parse(storedData) as string[]) : null,
  )
}

export async function setStoredDays(dayCollection: string[]) {
  return AsyncStorage.setItem(DAY_COLLECTION, JSON.stringify(dayCollection))
}
