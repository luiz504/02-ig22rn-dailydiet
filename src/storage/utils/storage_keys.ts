import { format, parse } from 'date-fns'

const FORMAT = 'yyyy/MM/dd'
export function storageDateKeyFormat(date: Date): string {
  return format(date, FORMAT)
}

export function storageKeyToDateParse(key: string) {
  return parse(key, FORMAT, new Date())
}
