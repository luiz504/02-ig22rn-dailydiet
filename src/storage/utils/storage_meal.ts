import AsyncStorage from '@react-native-async-storage/async-storage'
import { MEAL_COLLECTION } from '../commons'
import { Meal } from '~/models/Meal'

export type MealGroupByDay = {
  day: string
  meals: Meal[]
}

export async function getStoredMealsByDay(key: string) {
  return AsyncStorage.getItem(`${MEAL_COLLECTION}-${key}`).then((storedData) =>
    storedData ? (JSON.parse(storedData) as MealGroupByDay) : null,
  )
}

export async function setStoredMeals(key: string, meals: Meal[]) {
  const sortedMealsByHour = meals.sort(
    (a, b) => new Date(b.date).getTime() - new Date(a.date).getTime(),
  )
  const mealsCollection: MealGroupByDay = { day: key, meals: sortedMealsByHour }

  return AsyncStorage.setItem(
    `${MEAL_COLLECTION}-${key}`,
    JSON.stringify(mealsCollection),
  )
}

export async function getStoredMealsByDays(keys: string[]) {
  const keysCombined = keys.map((k) => `${MEAL_COLLECTION}-${k}`)
  const entries = await AsyncStorage.multiGet(keysCombined)
  return entries
    .map(([, value]) => (value ? JSON.parse(value) : null))
    .filter((value) => value) as MealGroupByDay[]
}

export async function deleteStoredMealCollection(key: string) {
  return AsyncStorage.removeItem(`${MEAL_COLLECTION}-${key}`)
}
