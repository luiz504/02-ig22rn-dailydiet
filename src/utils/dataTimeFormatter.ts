import * as Localization from 'expo-localization'

const userConfig = {
  uses24hoursClock:
    Localization?.getCalendars?.()?.[0]?.uses24hourClock ?? true,
  languageTag: Localization?.getLocales?.()?.[0]?.languageTag ?? 'en-US',
}

function formatDate(date: Date) {
  return new Intl.DateTimeFormat(userConfig.languageTag, {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(date)
}

function formatTime(time: Date) {
  return new Intl.DateTimeFormat(userConfig.languageTag, {
    hour: '2-digit',
    minute: '2-digit',
    hour12: !userConfig.uses24hoursClock,
  }).format(time)
}

function formatDateAndTime(date: Date | string) {
  const formattedFullDate = new Date(date).toLocaleDateString(
    userConfig.languageTag,
    {
      year: 'numeric',
      month: '2-digit',
      day: '2-digit',
      hour: 'numeric',
      minute: 'numeric',
    },
  )

  if (userConfig.languageTag.startsWith('en')) {
    return formattedFullDate.replace(',', ' at')
  }

  if (userConfig.languageTag.startsWith('pt')) {
    return formattedFullDate.replace(',', ' Ã s')
  }

  return formattedFullDate
}
export { userConfig, formatDate, formatTime, formatDateAndTime }
