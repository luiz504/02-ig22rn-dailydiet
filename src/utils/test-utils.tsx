import { FC, ReactNode } from 'react'
import { ThemeProvider } from 'styled-components/native'
import { render } from '@testing-library/react-native'

import { theme } from '~/styles'
import { NavigationContainer } from '@react-navigation/native'

const AllTheProviders: FC<{ children: ReactNode }> = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <NavigationContainer>{children}</NavigationContainer>
    </ThemeProvider>
  )
}

const ThemedWrapper: FC<{ children: ReactNode }> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender: typeof render = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

const renderThemed: typeof render = (ui, options) =>
  render(ui, { wrapper: ThemedWrapper, ...options })

const splitBorderColor = (borderColor: string) => ({
  borderTopColor: borderColor,
  borderRightColor: borderColor,
  borderBottomColor: borderColor,
  borderLeftColor: borderColor,
})

export * from '@testing-library/react-native'

export { customRender as render, renderThemed, splitBorderColor }
