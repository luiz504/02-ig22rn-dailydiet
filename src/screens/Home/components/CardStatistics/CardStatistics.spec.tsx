import { fireEvent, render, screen } from '~/utils/test-utils'
import { CardStatistics } from '.'
import { mockStatistics } from '../../mockMeals'
import { useNavigation } from '@react-navigation/native'

jest.mock('@react-navigation/native', () => {
  return {
    ...jest.requireActual('@react-navigation/native'),
    useNavigation: jest.fn(),
  }
})
describe('CardStatistics Component', () => {
  const buttonID = 'btn'

  const useNavigationSpy = () => {
    const navigate = jest.fn()
    jest
      .mocked(useNavigation)
      .mockReturnValue({ navigate, addListener: jest.fn() })
    return { navigate }
  }
  it('should render correctly with percentage >= 75', () => {
    render(
      <CardStatistics
        statistics={{ ...mockStatistics, inDietPercentage: 75.4 }}
        testID={buttonID}
      />,
    )
    const buttonElement = screen.getByTestId(buttonID)

    const arrowIconElement = screen.getByTestId('arrow-icon')
    const headingElement = screen.getByTestId('heading')
    const spanElement = screen.getByTestId('span')

    expect(buttonElement.props.style).toMatchSnapshot('btn-style')

    expect(headingElement).toHaveTextContent('75.40%')
    expect(headingElement.props.style).toMatchSnapshot('heading-style')

    expect(arrowIconElement.props.color).toMatchSnapshot(
      'arrow-icon-prop-inDiet',
    )

    expect(spanElement.props.style).toMatchSnapshot('text-style')
  })

  it('should render correctly with percentage < 75', () => {
    render(
      <CardStatistics
        statistics={{ ...mockStatistics, inDietPercentage: 40.111 }}
        testID={buttonID}
      />,
    )

    const arrowIconElement = screen.getByTestId('arrow-icon')

    expect(screen.getByTestId(buttonID).props.style).toMatchSnapshot(
      'btn-style-outDiet',
    )

    expect(screen.getByTestId('heading')).toHaveTextContent('40.11%')

    expect(arrowIconElement.props.color).toMatchSnapshot(
      'arrow-icon-prop-outDiet',
    )
  })

  it('should not render correctly when not provided statistics', () => {
    const { navigate } = useNavigationSpy()
    render(<CardStatistics statistics={null} testID={buttonID} />)

    const btnElement = screen.getByTestId(buttonID)
    expect(btnElement.props.style).toMatchSnapshot('btn-no-statistics')

    expect(screen.queryByTestId('arrow-icon')).toBeNull()
    expect(screen.getByTestId('heading')).toHaveTextContent('--')
    // Act

    fireEvent(btnElement, 'onPress')
    expect(navigate).not.toBeCalled()

    fireEvent(btnElement, 'onPress', '12312')
    expect(navigate).not.toBeCalled()
  })

  it('should navigate to Statistics screen when click on the Component', () => {
    const { navigate } = useNavigationSpy()
    render(<CardStatistics statistics={mockStatistics} testID={buttonID} />)

    const btnElement = screen.getByTestId(buttonID)

    fireEvent(btnElement, 'press')

    expect(navigate).toBeCalledTimes(1)
    expect(navigate).toHaveBeenCalledWith('statistics', {
      statistics: mockStatistics,
    })
  })
})
