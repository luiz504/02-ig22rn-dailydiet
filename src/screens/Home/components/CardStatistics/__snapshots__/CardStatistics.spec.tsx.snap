// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardStatistics Component should render correctly with percentage < 75: arrow-icon-prop-outDiet 1`] = `"#BF3B44"`;

exports[`CardStatistics Component should render correctly with percentage < 75: btn-style-outDiet 1`] = `
[
  {
    "borderBottomLeftRadius": 8,
    "borderBottomRightRadius": 8,
    "borderTopLeftRadius": 8,
    "borderTopRightRadius": 8,
    "paddingBottom": 20,
    "paddingLeft": 16,
    "paddingRight": 16,
    "paddingTop": 20,
    "position": "relative",
  },
  {
    "backgroundColor": "#F4E6E7",
  },
]
`;

exports[`CardStatistics Component should render correctly with percentage >= 75: arrow-icon-prop-inDiet 1`] = `"#639339"`;

exports[`CardStatistics Component should render correctly with percentage >= 75: btn-style 1`] = `
[
  {
    "borderBottomLeftRadius": 8,
    "borderBottomRightRadius": 8,
    "borderTopLeftRadius": 8,
    "borderTopRightRadius": 8,
    "paddingBottom": 20,
    "paddingLeft": 16,
    "paddingRight": 16,
    "paddingTop": 20,
    "position": "relative",
  },
  {
    "backgroundColor": "#E5F0DB",
  },
]
`;

exports[`CardStatistics Component should render correctly with percentage >= 75: heading-style 1`] = `
{
  "color": "#1B1D1E",
  "fontFamily": "Nunito-Bold",
  "fontSize": 32,
  "lineHeight": 41.6,
}
`;

exports[`CardStatistics Component should render correctly with percentage >= 75: text-style 1`] = `
{
  "color": "#333638",
  "fontFamily": "Nunito-Regular",
  "fontSize": 14,
  "lineHeight": 18.2,
}
`;
